src/Ext/Bullet/Trajectories/StraightTrajectory.cpp

```diff
@@ -121,100 +249,1037 @@ void StraightTrajectory::OnAIVelocity(BulletClass* pBullet, BulletVelocity* pSpe
 
 TrajectoryCheckReturnType StraightTrajectory::OnAITargetCoordCheck(BulletClass* pBullet)
 {
-	if (this->Type->PassThrough)
+	return TrajectoryCheckReturnType::SkipGameCheck; // No longer needed.
+}
+
+TrajectoryCheckReturnType StraightTrajectory::OnAITechnoCheck(BulletClass* pBullet, TechnoClass* pTechno)
+{
+	return TrajectoryCheckReturnType::SkipGameCheck; // Bypass game checks entirely.
+}
+
+void StraightTrajectory::PrepareForOpenFire(BulletClass* pBullet)
+{
+	const auto pType = this->Type;
+	double rotateAngle = 0.0;
+	const auto pTarget = pBullet->Target;
+	auto theTargetCoords = pBullet->TargetCoords;
+	auto theSourceCoords = pBullet->SourceCoords;
+
+	// TODO If I could calculate this before firing, perhaps it can solve the problem of one frame delay and not so correct turret orientation.
+	if (pType->LeadTimeCalculate && pTarget)
 	{
-		if (this->FirerZPosition > this->TargetZPosition && pBullet->Location.Z <= pBullet->TargetCoords.Z)
-			return TrajectoryCheckReturnType::Detonate; // Detonate projectile.
+		theTargetCoords = pTarget->GetCoords();
+		theSourceCoords = pBullet->Location;
+
+		if (theTargetCoords != this->LastTargetCoord)
+		{
+			const auto extraOffsetCoord = theTargetCoords - this->LastTargetCoord;
+			const auto targetSourceCoord = theSourceCoords - theTargetCoords;
+			const auto lastSourceCoord = theSourceCoords - this->LastTargetCoord;
+
+			const auto theDistanceSquared = targetSourceCoord.MagnitudeSquared();
+			const auto targetSpeedSquared = extraOffsetCoord.MagnitudeSquared();
+			const auto targetSpeed = sqrt(targetSpeedSquared);
+
+			const auto crossFactor = lastSourceCoord.CrossProduct(targetSourceCoord).MagnitudeSquared();
+			const auto verticalDistanceSquared = crossFactor / targetSpeedSquared;
+
+			const auto horizonDistanceSquared = theDistanceSquared - verticalDistanceSquared;
+			const auto horizonDistance = sqrt(horizonDistanceSquared);
+
+			const auto straightSpeedSquared = pType->Trajectory_Speed * pType->Trajectory_Speed;
+			const auto baseFactor = straightSpeedSquared - targetSpeedSquared;
+			const auto squareFactor = baseFactor * verticalDistanceSquared + straightSpeedSquared * horizonDistanceSquared;
+
+			if (squareFactor > 1e-10)
+			{
+				const auto minusFactor = -(horizonDistance * targetSpeed);
+				int travelTime = 0;
+
+				if (abs(baseFactor) < 1e-10)
```

Comment:
```suggestion
				if (std::abs(baseFactor) < 1e-10)
```
I know msvc won't confuse but add std:: just in case.